torch/csrc/jit/tensorexpr/schedule.cpp

```diff
@@ -399,6 +399,132 @@ class FunctionInliner : public IRMutator {
   std::unordered_set<const Var*> func_var_set_;
 };
 
+// Inlining for functions containing rand().  Since rand() is stateful we can't
+// simply inline it everywhere, or else we may generate new randoms where we
+// should us a previously generated one.  As a contrived example:
+//   %1 = rand()
+//   %2 = %1 + 1
+//   %3 = %1 - 1
+//   %4 = %2 - %3
+// Fully inlining this expr would, incorrectly, yield:
+//   %4 = (rand() + 1) - (rand() - 1)
+// when in fact the two uses of %1 should cancel.  To avoid this issue, we
+// instead generate:
+//   %4 = (let x = rand(); (x + 1) - (x - 1))
+//
+// The overall approach is to replace every rand() intrinsic with a newly
+// generated variable, and then bind those variables to rand() calls in the
+// body of the innermost for loop.
+class RandomInliner : public FunctionInliner {
+ public:
+  explicit RandomInliner(const std::vector<Function*>& funcs)
+      : FunctionInliner(funcs) {}
+
+  // Find the innermost for loop and bind any random variables in its body.
+  Stmt* mutate(const For* v) override {
```

Comment:
I am wondering whether we should respect other common C++ scopes: conditionals, etc. 

It seems wrong if we move an evaluation outside a if-then-else.